Analysis/src/GenParticle.cpp

```diff
@@ -0,0 +1,235 @@
+
+#include "h-tautau/Analysis/include/GenParticle.h"
+
+namespace analysis {
+
+GenParticle::GenParticle(const ntuple::Event& events, size_t n)
+    : index(n)
+{
+    if(n >= events.genParticles_p4.size())
+        throw std::runtime_error("events index is out of range.");
+
+    pdg = events.genParticles_pdg.at(n);
+    status = events.genParticles_status.at(n);
+    genStatusFlags = GenStatusFlags(events.genParticles_statusFlags.at(n));
+    momentum = events.genParticles_p4.at(n);
+    vertex = events.genParticles_vertex.at(n);
+}
+
+GenEvent::GenEvent(const ntuple::Event& event)
+{
+    genParticles.reserve(event.genParticles_p4.size());
+    for(size_t n = 0; n < event.genParticles_p4.size(); ++n)
+        genParticles.emplace_back(event,n);
+
+    for (size_t relIndex = 0; relIndex < event.genParticles_rel_mIndex.size(); relIndex++) {
+        int motherIndex = event.genParticles_rel_mIndex.at(relIndex);
+        int daughterIndex = event.genParticles_rel_pIndex.at(relIndex);
+
+        GenParticle* mother = &genParticles.at(static_cast<size_t>(motherIndex));
+        GenParticle* daughter = &genParticles.at(static_cast<size_t>(daughterIndex));
+
+        mother->daughters.push_back(daughter);
+        daughter->mothers.push_back(mother);
+    }
+
+    for(const GenParticle& genParticle : genParticles ){
+        if (!genParticle.mothers.size())
+            primaryParticles.insert(&genParticle);
+
+
+         if (genParticle.genStatusFlags.isPrompt())
+            particleCodeMap[std::abs(genParticle.pdg)].insert(&genParticle);
+    }
+}
+
+GenParticleSet GenEvent::GetParticles(int particle_pgd, bool requireIsLastCopy) const
+{
+    GenParticleSet results;
+    const ParticleCodeMap::const_iterator code_iter = particleCodeMap.find(particle_pgd);
+
+    if (code_iter != particleCodeMap.end()){
+        for (const GenParticle* particle : code_iter->second){
+            if(requireIsLastCopy && !particle->genStatusFlags.isLastCopy()) continue;
+            results.insert(particle);
+        }
+    }
+    return results;
+}
+
+void GenEvent::GetTypesParticles(const std::set<particles::ParticleCode::ParticleType>& type_names, const GenParticle* mother, GenParticleSet& result) const
+{
+    if(mother->genStatusFlags.isPrompt() && mother->genStatusFlags.isLastCopy() && particle_types.count(mother->pdg)
+        && type_names.count(particle_types.at(mother->pdg))) {
+//        if(shared && areParented(mother,mother->daughters.at(0)) && areParented(mother,mother->daughters.at(1)) )
+        result.insert(mother);
+    } else {
+        for(const GenParticle* daughter : mother->daughters)
+            GetTypesParticles(type_names, daughter, result);
+    }
+}
+
+bool GenEvent::areParented(const GenParticle* daughter, const GenParticle* mother) const
+{
+    for (size_t i = 0; i < daughter->mothers.size(); i++) {
+        const GenParticle* mother_ = daughter->mothers.at(i);
+        if(mother_ == mother)
```

Comment:
put both conditions in one line, i.e. `if(mother_ == mother || areParented(mother_, mother) return true;`